### PR TITLE
 feat(oauth): cherry-pick from v0.8 to v0.9

### PR DESCRIPTION
**Pull request by moshuipan**
**What this PR does / why we need it**:

**Which issue(s) this PR is related to** *(optional, link to 3rd issue(s))*:

Fixes #

Reference to #

**Special notes for your reviewer**:

/cc @zhujian

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/server/apis/v1/handler/oauth.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+
+	// "github.com/caicloud/cyclone/cmd/worker/options"
```
Comment (zhujian7) [inline pkg/server/apis/v1/handler/oauth.go]: please delete unused code

File: pkg/server/biz/scm/gitlab/gitlab.go [hunk 2]
```diff
@@ -92,7 +92,8 @@ func NewGitlab(scmCfg *v1alpha1.SCMSource) (scm.Provider, error) {
 // newGitlabV4Client news Gitlab v4 client by token. If username is empty, use private-token instead of oauth2.0 token.
 func newGitlabV4Client(server, username, token string) (*v4.Client, error) {
 	var client *v4.Client
-	if len(username) == 0 {
+	if ensureTokenType(server, token) == PrivateToken {
+		// if len(username) == 0 {
```
Comment (zhujian7) [inline pkg/server/biz/scm/gitlab/gitlab.go]: delete?

File: pkg/server/apis/v1/handler/oauth.go (hunk 1 above)
Comment (moshuipan) [reply pkg/server/apis/v1/handler/oauth.go]: deleted

**Comment by moshuipan** [comment]
Reference to #https://github.com/caicloud/service/issues/762

File: pkg/server/apis/v1/handler/oauth.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+
+	api "github.com/caicloud/cyclone/pkg/server/apis/v1"
```
Comment (zhujian7) [inline pkg/server/apis/v1/handler/oauth.go]: please format it

**Pull request merged**